components/layout_2020/flow/mod.rs

```diff
@@ -433,80 +430,70 @@ fn layout_in_flow_non_replaced_block_level<'a>(
         min_box_size.block == Length::zero() &&
         pb.block_end == Length::zero() &&
         block_level_kind == BlockLevelKind::SameFormattingContextBlock;
-    let mut nested_positioning_context = PositioningContext { abspos: Vec::new() };
-    let mut flow_layout = layout_contents(
-        if style.get_box().position == Position::Relative {
-            &mut nested_positioning_context
-        } else {
-            positioning_context
-        },
-        &containing_block_for_children,
-        this_start_margin_can_collapse_with_children,
-    );
-    let mut block_margins_collapsed_with_children = CollapsedBlockMargins::from_margin(&margin);
-    if this_start_margin_can_collapse_with_children.0 {
-        block_margins_collapsed_with_children
-            .start
-            .adjoin_assign(&flow_layout.collapsible_margins_in_children.start);
-        if flow_layout
-            .collapsible_margins_in_children
-            .collapsed_through
-        {
+
+    positioning_context.for_maybe_position_relative(layout_context, style, |positioning_context| {
```

Comment:
I understand the logic but I'm not sure I like how big this closure is and how it drifts everything even more to the right. Do you think it could be its own function, living in the module next to this one?